nrf-mpsl/src/flash.rs

```diff
@@ -0,0 +1,496 @@
+                if let FlashOp::None = state.operation {
+                    let mut session_id: u8 = 0;
+                    let ret = unsafe { raw::mpsl_timeslot_session_open(Some(timeslot_session_callback), (&mut session_id) as *mut _) };
+                    match RetVal::from(ret).to_result() {
+                        Ok(_) => {
+                            Poll::Ready(Ok(session_id))
+                        }
+                        Err(Error::ENOMEM) => {
+                            state.waker.register(cx.waker());
+                            Poll::Pending
+                        }
+                        Err(e) => {
+                            Poll::Ready(Err(e))
+                        }
+                    }
+                } else {
+                    Poll::Pending
+                }
+            })
+        }).await?;
+
+        let _drop = OnDrop::new(|| {
+            STATE.with_inner(|state| {
+                state.operation = FlashOp::None;
+            });
+            let _ = unsafe { raw::mpsl_timeslot_session_close(session_id) };
+        });
+
+        // info!("Requesting timeslot");
+        STATE.with_inner(|state| {
+            state.result = None;
+            state.operation = op;
+            state.slot_duration_us = slot_duration_us;
+            state.timeslot_request.request_type = raw::MPSL_TIMESLOT_REQ_TYPE_EARLIEST as u8;
+            state.timeslot_request.params.earliest = raw::mpsl_timeslot_request_earliest_t {
+                hfclk: raw::MPSL_TIMESLOT_HFCLK_CFG_NO_GUARANTEE as u8,
+                priority: raw::MPSL_TIMESLOT_PRIORITY_NORMAL as u8,
+                length_us: slot_duration_us + TIMESLOT_SLACK_US,
+                timeout_us: TIMESLOT_TIMEOUT_PRIORITY_NORMAL_US,
+            };
+            let ret = unsafe { raw::mpsl_timeslot_request(session_id, &state.timeslot_request) };
+            RetVal::from(ret).to_result()
+        })?;
+        // info!("Timeslot requested");
+
+        poll_fn(|cx| {
+            STATE.with_inner(|state| {
+                match state.result.take() {
+                    Some(result) => {
+                        state.operation = FlashOp::None;
+                        Poll::Ready(result)
+                    }
+                    None => {
+                        state.waker.register(cx.waker());
+                        Poll::Pending
+                    }
+                }
+            })
+        })
+        .await?;
+
+        // No need to run the drop handler now
+        _drop.defuse();
+
+        unsafe {
+            // info!("MPSL closing session {}", session_id);
+            let ret = raw::mpsl_timeslot_session_close(session_id);
+            RetVal::from(ret).to_result()?;
+            // info!("MPSL closed session {}", session_id);
+        }
+
+        Ok(())
+    }
+
+    pub async fn erase(&mut self, from: u32, to: u32) -> Result<(), FlashError> {
+        if to < from || to as usize > FLASH_SIZE {
```

Comment:
Is a zero-length erase valid? If so should it erase anything?

Reply:
I modified the operations to deal with this case, it should not erase anything (to is exclusive). These guards are the same as in nrf-softdevice and embassy-nrf nvmc, so I think it should deal with this appropriately now.